### PR TITLE
Support parallelized downloading in load_dataset with Spark

### PR DESCRIPTION
**Pull request by es94129**
As proposed in https://github.com/huggingface/datasets/issues/5798, this adds support to parallelized downloading in `load_dataset` with Spark, which can speed up the process by distributing the workload to worker nodes.

Parallelizing dataset processing is not supported in this PR.

### CODE REVIEW

File: src/datasets/utils/spark_utils.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import os
+import uuid
+
+import pyspark
+
+
+def validate_cache_dir(cache_dir, spark_function_name):
```
Comment (maddiedawson) [inline src/datasets/utils/spark_utils.py:7]: Let's add an option to pass in spark so that Spark's self._spark can be reused?

Comment (maddiedawson) [reply src/datasets/utils/spark_utils.py:7]: Also let's add a comment on this function that also describes what `spark_function_name` is being used for (error message)

**Review by maddiedawson — COMMENTED** [review]
Looks good overall!

File: src/datasets/utils/py_utils.py [hunk 2]
```diff
@@ -439,11 +440,29 @@ def map_nested(
...
+            import pyspark
+
+            spark = pyspark.sql.SparkSession.builder.getOrCreate()
+            mapped = (
+                spark.sparkContext.parallelize(iterable)
+                .map(lambda obj: _single_map_nested((function, obj, types, None, True, None)))
+                .collect()
+            )
+
+        except ImportError:
```
Comment (lu-wang-dl) [inline src/datasets/utils/py_utils.py:460]: We may need to provide some more details for the exceptions. If the import failed, we need to give a warning saying something like "pyspark module is not found. Downloading in sequentially". If spark session creation fails, we need to give corresponding messages.

File: src/datasets/utils/spark_utils.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import os
+import uuid
+from typing import Optional
+
+import pyspark
```
Comment (lu-wang-dl) [inline src/datasets/utils/spark_utils.py:5]: We need to put this import into try catch. pyspark is not a hard dependencies for datasets.

**Comment by es94129** [comment]
Hi @lhoestq or other maintainers, this is ready for review, could you please take a look?

**Pull request closed without merging**